src/node.cc

```diff
@@ -1138,14 +1146,59 @@ void PromiseRejectCallback(PromiseRejectMessage message) {
   callback->Call(process, arraysize(args), args);
 }
 
+void OnPromiseGC(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsObject());
+  Local<Object> promise = args[0].As<Object>();
+
+  TrackPromise::New(env->isolate(), promise);
+
+  HandleScope scope(env->isolate());
```

Comment:
JS -> C++ callbacks have an implicit HandleScope, you don't need to create one yourself.